### PR TITLE
Draft of async API

### PR DESCRIPTION
**Pull request by lutovich**
PR adds initial version of async API for direct (only `bolt://` scheme, not `bolt+routing://`) driver.

Main API contact points:
 * `Response` interface that is a `Future` one can subscribe to with `ResponseListener`
 * `StatementResultCursor` which is an async counterpart of `StatementResult`; it is basically an async iterator with `#hasNext()` being `Response<Boolean> fetchAsync()` and `#next()` being `Record current()`
 * various overloads or `Session#runAsync()` to run queries asynchronously in auto-commit mode
 * `Session#beginTransactionAsync()` to start a transaction asynchronously
 * various overloads or `Transaction#runAsync()` to run queries asynchronously in the transaction
 * `Transaction#commitAsync()` and `Transaction#rollbackAsync()` to finish transaction with commit or rollback
 * `Session#closeAsync()` to close session asynchronously and release all resources

Implementation notes:
 * Netty is used internally for async IO; it is now specified as a dependency with scope `compile` but will eventually be shaded into the driver artifact
 * All internal futures & promises are built on top of corresponding Netty classes; all callbacks and future transformations execute in Netty event loop threads
 * Channels are pooled in a `FixedChannelPool` which adds two new config settings: `maxConnectionPoolSize` and `connectionAcquisitionTimeoutMillis`
 * Javadocs are missing right now
 * Packages will need some reshuffling
 * Simple back-pressure is available using channel auto-read functionality but it's turned off by default
 * Async API lives in parallel to sync API and uses separate network connections, pools and everything; this is temporary and eventually sync will be implemented on top of async

Not part of this PR:
 * Transaction functions with retries
 * Routing driver
 * Netty logging does not use driver logging
 * Protocol level back-pressure
 * Good test coverage

Related to https://github.com/neo4j/neo4j-java-driver/issues/102

### CODE REVIEW

**Comment by zhenlineo** [comment]
LG, let's merge!

BTW, Should we merge this into a separate branch? or should we merge into main branch?

File: driver/src/main/java/org/neo4j/driver/internal/ConnectionSettings.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
  */
 public class ConnectionSettings
 {
-    private static final String DEFAULT_USER_AGENT = format( "neo4j-java/%s", driverVersion() );
+    public static final String DEFAULT_USER_AGENT = format( "neo4j-java/%s", driverVersion() );
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/ConnectionSettings.java]: Very small thing but is this being used elsewhere? Shall we set it as private again?

File: driver/src/main/java/org/neo4j/driver/internal/ExplicitTransaction.java [hunk 2]
```diff
@@ -57,28 +70,68 @@
...
+        this.asyncConnection = asyncConnection;
+        this.session = session;
+        this.resourcesHandler = SessionResourcesHandler.NO_OP;
+    }
+
+    public void begin( Bookmark initialBookmark )
+    {
+        Map<String,Value> parameters = initialBookmark.asBeginTransactionParameters();
+
+        connection.run( "BEGIN", parameters, NoOpResponseHandler.INSTANCE );
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/ExplicitTransaction.java]: Do these "BEGIN", "COMMIT" and "ROLLBACK" strings  deserve their constants?

**Review by ali-ince — APPROVED** [review]
Looks really good! Only a couple of comments (even can be ignored :)).

**Comment by lutovich** [comment]
We decided to keep async API in a separate branch for now. It will get merged into 1.5 a bit later.
I'll close this PR now. Please track progress in https://github.com/neo4j/neo4j-java-driver/tree/1.5-async.

**Pull request closed without merging**